downloadDirectory/src/main/java/org/reactome/release/downloadDirectory/GenerateGOAnnotationFile/BiologicalProcessAnnotationBuilder.java

```diff
@@ -0,0 +1,122 @@
+
+        for (GKInstance proteinInst : proteinInstances) {
+            GKInstance referenceEntityInst = (GKInstance) proteinInst.getAttributeValue(ReactomeJavaConstants.referenceEntity);
+            GKInstance speciesInst = (GKInstance) proteinInst.getAttributeValue(ReactomeJavaConstants.species);
+            // Check if the protein has any disqualifying attributes.
+            boolean validProtein = GOAGeneratorUtilities.validateProtein(referenceEntityInst, speciesInst);
+            if (validProtein) {
+                String taxonIdentifier = ((GKInstance) speciesInst.getAttributeValue(ReactomeJavaConstants.crossReference)).getAttributeValue(ReactomeJavaConstants.identifier).toString();
+                // TODO: Rule check
+                if (!microbialSpeciesToExclude.contains(taxonIdentifier)) {
+                    getGOBiologicalProcessLine(referenceEntityInst, reactionInst, taxonIdentifier);
+                }
+            }
+        }
+    }
+
+    // Before creating GOA line for BP annotations, the reaction in question needs to be checked for the existance of a 'goBiologicalProcess' attribute. If there is none
+    // than the instance's 'hasEvent' referrals are checked for any.
+    private static List<String> getGOBiologicalProcessLine(GKInstance referenceEntityInst, GKInstance reactionInst, String taxonIdentifier) throws Exception {
+        List<String> goAnnotationLines = new ArrayList<>();
+        for (Map<String, String> biologicalProcessAccession : getGOBiologicalProcessAccessions(reactionInst, 0)) {
+            String goaLine = GOAGeneratorUtilities.generateGOALine(referenceEntityInst, "P", biologicalProcessAccession.get("accession"), biologicalProcessAccession.get("event"), "TAS", taxonIdentifier);
+            GOAGeneratorUtilities.assignDateForGOALine(reactionInst, goaLine);
+            goAnnotationLines.add(goaLine);
+        }
+        return goAnnotationLines;
+    }
+
+    // This method checks for a populated 'goBiologicalProcess' attribute in the incoming instance. If there are none and the max recursion has been reached,
+    // it's 'hasEvent' referral is checked for it. Once finding it, it returns the 'accession' and 'identifier' for each one, which will be used to generate a GOA line.
+    private static List<Map<String, String>> getGOBiologicalProcessAccessions(GKInstance eventInst, int recursion) throws Exception {
+        List<Map<String, String>> goBiologicalProcessAccessions = new ArrayList<>();
+        if (recursion <= MAX_RECURSION_LEVEL) {
+            Collection<GKInstance> goBiologicalProcessInstances = eventInst.getAttributeValuesList(ReactomeJavaConstants.goBiologicalProcess);
+            if (goBiologicalProcessInstances.size() > 0) {
+                for (GKInstance goBiologicalProcessInst : goBiologicalProcessInstances) {
+                    if (!goBiologicalProcessInst.getAttributeValue(ReactomeJavaConstants.accession).toString().equals(PROTEIN_BINDING_ANNOTATION)) {
```

Comment:
This condition is a little complicated.  Could it extracted a function and read `if(!proteinBindingAccession(goBiologicalProcessInst))`?

Reply:
Yea I made a boolean method that should be a bit more readable